File: Algorithm/e-olymp/4004_matrix.java

import java.util.Scanner;

public class Solution{
	
	private final static boolean IS_DEBUG = true;
	private final static int NUMBER_OF_VERTICES = 50;
	private static final String IDS_INPUT_FILE_NAME = "input.txt";
	
	private boolean graph[][];
	private boolean visted[];
	
	int vertexCount = -1;
	
	public Solution() {
		graph = new boolean[ NUMBER_OF_VERTICES ][ NUMBER_OF_VERTICES ];
		visted = new boolean[ NUMBER_OF_VERTICES ];
	}
	
	public Scanner getScanner() {
		if ( IS_DEBUG ){
			return new Scanner(getClass().getResourceAsStream(IDS_INPUT_FILE_NAME));
		} else {
			return new Scanner(System.in);
		}
	}

	public static void main(String[] args) {
		Solution solution = new Solution();		
		System.out.print( solution.getAnswer() );
	}

	public int getAnswer(){
		
		Scanner sc = getScanner();
		
		if (sc.hasNextInt()){
			vertexCount = sc.nextInt();		
		}
		
		for ( int y = 0; y < vertexCount; ++y ){
			for ( int x = 0; x < vertexCount; ++x ){
				graph[y][x] = sc.nextInt() == 1;
			}
			visted[y] = false;
		}
		
		for ( int v = 0; v < vertexCount; ++v ){
			if ( dfs( v ) ){
				return 1;
			}
		}
		
		return 0;
	}

	private boolean dfs(int vFrom) {
		visted[ vFrom ] = true;
		for ( int vTo = 0; vTo < vertexCount; ++vTo ){
			if ( graph[ vFrom ][ vTo ] ){
				if ( visted[vTo] ){
					return true;
				}

				if ( dfs( vTo )){
					return true;
				}
			}
		}
		visted[ vFrom ] = false;
		return false;
	}

}

File: Algorithm/e-olymp/4004_matrix.cpp
#include "stdio.h"
#include <iostream>

using namespace std;

#define SIZE 50

int n;
bool graph [SIZE][SIZE];
bool visited [SIZE][SIZE];

void fill(){
    cin >> n;
    for ( int vF = 0; vF < n; ++vF ){
        for ( int vT = 0; vT < n; ++vT ){
            cin >> graph[vF][vT];
            visited[vF][vT] = 0;
        }
    }
}

bool dfs( int vF ){
    for ( int vT = 0; vT < n; ++vT ){
        if ( graph[vF][vT] == 0 ){
            continue;
        }

        if ( visited[vF][vT] ){
            return true;
        }

        visited[vF][vT] = 1;
        if ( dfs( vT ) ){
            return true;
        }
        visited[vF][vT] = 0;
    }
    return false;
}

bool solution(){
    for ( int vF = 0; vF < n; ++vF ){
        if ( dfs(vF) ){
            return 1;
        }
	}
	return 0;
}

int main(int argc, char** argv)
{
	//freopen("input.txt", "r", stdin);

	fill();
    cout << solution();
    return 0;
}

